### PR TITLE
[Previous draft] Read/write FASTQ as unmapped SAM

### PR DESCRIPTION
**Pull request by jmarshall**
For entertainment value and comparison to PR #1156 (i.e. to see if there are any aspects here that would be useful in the other PR), here is the branch that has been gathering dust since [this conversation](https://twitter.com/jomarnz/status/788228002270285824).

### CODE REVIEW

**Comment by jkbonfield** [comment]
It maybe would have been useful if you could have put this in as a PR in 2016!  What other useful snippets have you been sitting on? ;-)

There are some useful things in there though that could be merged over.  I did initially start with my own fastq parser, mainly because getting kseq to work was such a huge uphill struggle!  I did eventually beat it into submission though so went via that route.

Writing is the bit I haven't got yet (and arguably samtools fastq covers this from the application side, although not the library side), but I was dissuaded from this course of action by others.  Do people think writing from the library is useful?  My instincts are yes.

File: htslib/hts.h [hunk 1]
```diff
@@ -457,14 +457,15 @@ char *hts_format_description(const htsFormat *format);
   @param fn       The file name or "-" for stdin/stdout. For indexed files
                   with a non-standard naming, the file name can include the
                   name of the index file delimited with HTS_IDX_DELIM
-  @param mode     Mode matching / [rwa][bceguxz0-9]* /
+  @param mode     Mode matching / [rwa][bcefguxz0-9]* /
```
Comment (jkbonfield) [inline htslib/hts.h:460]: What's the x and e in this string?  I can't see any support for these in the code, and they're undocumented below.  Some legacy thing?

Comment (jmarshall) [reply htslib/hts.h:460]: They are passed through to `hopen()` and interpreted there (see #245), and actually are documented in the `hts_open()` blurb — look further below.

Comment (jkbonfield) [reply htslib/hts.h:460]: I should have opened the file to look.  :)

I saw them documented in alphabetical order and github trimmed the surrounding text at just the wrong spot.  Duh!  Sorry and thanks for the explanation.

I'll be adding bits of this as I now have the go ahead for doing the writing support too, which will start from this.  Cheers.

**Comment by jkbonfield** [comment]
Obsoleted by #1156

**Pull request closed without merging**